### ISSUE TITLE
a suitable version of java could not be found

### ISSUE THREAD

**Issue by dmaziuk** [mguessan/davmail#160]
64-bit 5.5.1-3299, windows 10, corretto 11 jdk

**Comment by dmaziuk** [comment]
PS JAVA_HOME is set, `java -version` in powershell or git bash returns
```
openjdk version "11.0.11" 2021-04-20 LTS
OpenJDK Runtime Environment Corretto-11.0.11.9.1 (build 11.0.11+9-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.11.9.1 (build 11.0.11+9-LTS, mixed mode)
```
and running `davmail64.exe` from command line prints
```
[info] Configured vm.location: (null)
```

**Comment by ringosham** [comment]
I have a different log. Davmail assumes the jvm to be in its current working directory rather than JAVA_HOME?
```
[info] Module Name: C:\Program Files\DavMail\davmail64.exe
[info] Module INI: C:\Program Files\DavMail\davmail64.ini
[info] Module Dir: C:\Program Files\DavMail\
[info] INI Dir: C:\Program Files\DavMail\
[info] Configured vm.location: ./jre/bin/server/jvm.dll
[info] vm.location item not found: ./jre/bin/server/jvm.dll
 [err] A suitable version of Java could not be found on your system, please install Java (64-bit) from http://java.com/en/download/manual.jsp
```

**Comment by s-p-turner** [comment]
I also encountered this issue. One workaround is to create a davmail64.ini file in C:\Program Files\DavMail containing a vm.location parameter to your jvm.dll, e.g.:

`vm.location=C:\Program Files\Java\jre1.8.0_291\bin\server\jvm.dll`

**Comment by ringosham** [comment]
@s-p-turner This works.

For better usability they should automatically generate this file, or at least tell us what we should do in a message box.

**Comment by dmaziuk** [comment]
well... $JAVA_HOME has been a de-facto standard only since, like, the 90s... but yes: so it installing a "default" config file with comments and instructions.

**Comment by s-p-turner** [comment]
From what I can gather, the 64-bit Java installer doesn't add anything to the environment that would allow DavMail to pick up the path to the jvm.dll file. The 32-bit Java installer does - it adds "C:\Program Files (x86)\Common Files\Oracle\Java\javapath" to the "Path" variable, and "javapath" is maintained as a shortcut to the directory containing the actual version installed.

To automatically generate the davmail64.ini file during installation in the absence of any environmental setup by Java-64, the DavMail installer would have to go searching for the 64-bit jvm.dll in "C:\Program Files\Java". But even if it successfully did that and created the .ini file, the path to the 64-bit jvm.dll would always contain the revision of Java installed, and the entry in the .ini file would need to be manually updated if/when you updated Java.

I think the 64-bit Java installer might be to blame here, rather than DavMail itself. But I do agree that the message contained in the error dialog could be rather more descriptive!

**Comment by vam103** [comment]
Well I tried all of the methods here with my AdoptOpenJDK JRE 16 on Windows 10 and version 6 is still saying it can not find java. Which is weird since version 5 can and runs quite well. AdoptOpenJDK even set the JAVA_HOME variable for the 64-bit version. i have tried with multiple versions of AdoptOpenJDK and it seems not to like it. Meanwhile the .jar works for version 6 with the same setup.

**Comment by s-p-turner** [comment]
I don't think davmail64.exe checks for the JAVA_HOME variable being set.

**Comment by JamieKitson** [comment]
The odd thing is that this is a change in 6.0. 5.5 worked fine for me, 6.0 needed the ini file fix.

**Comment by dmaziuk** [comment]
> 
> 
> I don't think davmail64.exe checks for the JAVA_HOME variable being set.

It should. 

`$PATH` is for executables so the 64-bit Corretto installer adds `bin` where `java.exe` is. It does not add `bin/server`, nor should it: there are no executables there. That, and java versioning, is why `davmail64.exe` should be looking in `${JAVA_HOME}/bin/server` and not in `$PATH` or anywhere else.

Also the `default ini file` is not something to generate during install, it's a boilerplate config file that has all settings commented out (and hopefully commented). So the users familiar with the practice can trace `vm.location` to it without making a bug report.

Anyway, 6.0.0 works with the `ini` fix. Yay!

**Comment by s-p-turner** [comment]
> 
> 
> > I don't think davmail64.exe checks for the JAVA_HOME variable being set.
> 
> It should.

Agreed. I was just pointing out that I don't think it currently does. I extracted all the strings from the executable and JAVA_HOME just isn't in there.

> Anyway, 6.0.0 works with the `ini` fix. Yay!

:-)

**Comment by vam103** [comment]
Well the jvm is found with the ini file. Probably the openj9 jre gets the error
`JVMJ9VM007E Command-line option unrecognised: java.base/sun.net.www.protocol.https=ALL-UNNAMED`
Seems to be down to a change that added
`[info] vmarg.2=--add-exports
[info] vmarg.3=java.base/sun.net.www.protocol.https=ALL-UNNAMED`
Anyway more digging to do.

**Comment by s-p-turner** [comment]
> 
> 
> try running `java -jar davmail.jar` instead of `.exe`, see what happens

Executing davmail.jar did work for me. The issue was always with executing the .exe. But to use this as a definitive workaround meant modifying shortcuts etc. Therefore the davmail64.ini file workaround that I originally suggested was more suitable for me.

**Comment by dmaziuk** [comment]
`.ini` is in here: `https://github.com/mguessan/davmail/tree/master/src/winrun4j/` but it remains to be seen if a) overriding `vmarg.N` with an empty string in the `.ini` will work and b) that particular setting is actually needed. By "see what happens" I meant the later: see if it runs without.

**Comment by lihatt** [comment]
Same situation here: Win 10; Davmail 6.0.0-3375 did not detect Java 8u291; the .ini file workaround worked; as JamieKitson said, 5.5 used to work fine without this extra intervention.

**Comment by Reiner030** [comment]
I found the cause for this issue in my setup.
Checking env path with `echo %PATH%`  in cmd console I saw that I had several old (and one active) x86 and 64 binary java path entries so first match wins which is in old setups like my one mainly the x86 version path.

After removing this in system environment settings => environments variables => system variables it works as expected.

**Comment by lihatt** [comment]
I just updated Java and was confused for a second why this broke DavMail again, until I remembered I had hard-coded the path to the previous version in an ini file as described above. I tried the solution in Reiner030's comment about removing references to x86 Java in the Windows path but it didn't have any effect.

Is there any hope this will get fixed?

**Comment by Reiner030** [comment]
I use also DavMail 6.0.0-3375 but
```
$ java -version
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)
```
which I fetched from https://adoptopenjdk.net/ last time.

**Comment by marcelser** [comment]
Same for me cleaning up path didn't have any effect. It still doesn't find the java version. Only ini file fix solved it for me. But as others have pointed out. If you update your java version you have to remember to update the .ini file.

This issue now open since 9 months and still no one has fixed it. It's not even assigned. When will one of the devs pick it up?

**Comment by konzertnr9** [comment]
I'm using
`vm.location=%JAVA_HOME%\bin\server\jvm.dll`

**Comment by vam103** [comment]
Since I updated to version 6.0.1-3390 I have not needed the ini file while running IBM Semeru Open runtime 17 (IBM's version of AdoptOpenJDK with openj9) and it seems to be working on multiple machines running Windows 10.

**Comment by s-p-turner** [comment]
> I'm using `vm.location=%JAVA_HOME%\bin\server\jvm.dll`

JAVA_HOME isn't populated in my setup - the variable is empty. I guess I could always populate it myself, add it to my environment and use it for the vm.location setting the .ini file. However, I'd still have to remember to change JAVA_HOME manually each time I updated Java. Therefore it would be no different to my current workaround of manaully editing the .ini file.

Is the JAVA_HOME environment variable supposed to be set up by the Java install?

**Comment by konzertnr9** [comment]
It's just my practice to check JAVA_HOME and update it where necessary when I update my JVM.

**Comment by marcelser** [comment]
> 

That's not the question. The question is: should java installers set JAVA_HOME properly and davmail pick it up for locating java or not. If you're installing java from some archive then it's on your hands of course.

**Comment by konzertnr9** [comment]
The Oracle updator does not populate JAVA_HOME.

**Comment by lihatt** [comment]
Yes, it's really rather irritating that I have to manually change this ini file every time Java is updated.

**Comment by marcelser** [comment]
Unfortunately it seems the davmail devs don't care at all about this problem. This bug was opened 27th May 2021 which is now more then 1 year ago. No one is assigned to it it's not even confirmed. And yes it's a pain in the ass to always update the ini file but it seems it has lowest priority and will probably lets face it never be addressed

**Comment by Anguel** [comment]
I have now written a small Windows batch script to automatically set JAVA_HOME. You need to adapt the **java_home_wildcard** starting path part (that will not change on Java updates) to match your Java installs. This starting path part must have the * wildcard at its end.
This way Windows can change directory to the first folder it finds matching the wildcard which will then result in a cd to the latest installed Java directory. Of course your java installers/updaters must not leave previous versions in that folder, otherwise it will not work.
The current directory can be now used by the script as the %cd% variable.

My example **auto_set_java_home.bat** batch below is for the 64 bit version of AdoptOpenJDK 11. The **wildcard path should be adapted for other Java providers and versions**:

```
@echo off
:: Set the java_home_wildcard to point to the name your Java folder starts with, followed by the wildcard asterisk
set java_home_wildcard="C:\Program Files\Eclipse Adoptium\jre-11*"
:: This batch script will then permanently set JAVA_HOME to the first matching folder, it must be run as Administrator!
echo This batch script will set JAVA_HOME permanently, run it as Administrator!
cd %java_home_wildcard%
echo Setting JAVA_HOME to %cd%
setx /M JAVA_HOME "%cd%"
pause
```

Additionally, the **davmail64.ini** must be created in C:\Program Files\DavMail and contain:

`vm.location=%JAVA_HOME%\bin\server\jvm.dll`

I decided to use this JAVA_HOME approach instead of writing the full Java path directly to the .ini file, you never know if DavMail will not add something else to the .ini at some point that may get overwritten by the script...

Whenever DavMail does not start due to a Java update anymore you just rerun the batch as admin and it updates JAVA_HOME.

**Comment by lihatt** [comment]
Feh, I had some hope there with that batch script, but even changing the ini file to the string you specify, it doesn't seem to work. Windows reports the setting was changed but Davmail's still not seeing it.

**Comment by Carterpersall** [comment]
@lihatt I use OpenJDK and don't encounter the issue you have as all updates within a major version (Java 8, Java 16, etc.) have the same folder structure. I would recommend most people use OpenJDK because Oracle. I've used both Adoptium (https://adoptium.net/) and Zulu (https://www.azul.com/downloads/) in the past with no issues.

**Comment by konzertnr9** [comment]
@lihatt Have you restarted your computer after running the script?
My PC with Windows 11 installed requires a restart after JAVA_HOME has been set / updated.

**Comment by Anguel** [comment]
@Carterpersall I use Ninite Pro for auto-updating Adoptium and as expected one of the latest updates broke the path, so I had to rerun my script and now it works again. This maybe does not happen with pure Adoptium updates.

**Comment by bbrown-spsu** [comment]
> I found the cause for this issue in my setup. Checking env path with `echo %PATH%` in cmd console I saw that I had several old (and one active) x86 and 64 binary java path entries so first match wins which is in old setups like my one mainly the x86 version path.
> 
> After removing this in system environment settings => environments variables => system variables it works as expected.

That fixed it for me as well.  Thank you.

**Comment by mguessan** [comment]
Eventually fixed by providing a package with custom JRE included.

Thanks for your feedback.
